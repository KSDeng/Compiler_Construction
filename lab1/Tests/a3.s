.data
_prompt: .asciiz "Enter an integer:"
_ret: .asciiz "\n"
.globl main
.text

read:
  li $v0, 4
  la $a0, _prompt
  syscall
  li $v0, 5
  syscall
  jr $ra

write:
  li $v0, 1
  syscall
  li $v0, 4
  la $a0, _ret
  syscall
  move $v0, $0
  jr $ra


main:
  li $t0, 0
  addi $sp, $sp, -4
  sw $ra, 0($sp)
  jal read
  lw $ra, 0($sp)
  addi $sp, $sp, 4
  move $t1, $v0
  move $t2, $t1
  addi $sp, $sp, -4
  sw $ra, 0($sp)
  jal read
  lw $ra, 0($sp)
  addi $sp, $sp, 4
  move $t3, $v0
  move $t4, $t3
  bgt $t2, $t4, label0
  j label1
label0:
  move $t5, $t2
  j label2
label1:
  move $t5, $t4
label2:
  li $t6, 0
  beq $t0, $t6, label4
  j label5
label4:
  div $t5, $t2
  mflo $t7
  mul $s0, $t7, $t2
  beq $t5, $s0, label6
  j label7
label6:
  div $t5, $t4
  mflo $s1
  mul $s2, $s1, $t4
  beq $t5, $s2, label9
  j label10
label9:
  move $s3, $t5
  li $t0, 1
  j label2
label10:
  addi $s4, $t5, 1
  move $t5, $s4
  j label2
label7:
  addi $s5, $t5, 1
  move $t5, $s5
  j label2
label5:
  move $s6, $s3
  move $a0, $s6
  addi $sp, $sp, -4
  sw $ra, 0($sp)
  jal write
  lw $ra, 0($sp)
  addi $sp, $sp, 4
  move $v0, $0
  jr $ra
