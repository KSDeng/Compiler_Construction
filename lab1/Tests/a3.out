Program (1)
  ExtDefList (1)
    ExtDef (1)
      Specifier (1)
        TYPE: int
      FunDec (1)
        ID: main
        LP
        RP
      CompSt (1)
        LC
        DefList (2)
          Def (2)
            Specifier (2)
              TYPE: int
            DecList (2)
              Dec (2)
                VarDec (2)
                  ID: a
              COMMA
              DecList (2)
                Dec (2)
                  VarDec (2)
                    ID: b
            SEMI
          DefList (3)
            Def (3)
              Specifier (3)
                TYPE: int
              DecList (3)
                Dec (3)
                  VarDec (3)
                    ID: result
              SEMI
            DefList (4)
              Def (4)
                Specifier (4)
                  TYPE: int
                DecList (4)
                  Dec (4)
                    VarDec (4)
                      ID: start
                SEMI
              DefList (5)
                Def (5)
                  Specifier (5)
                    TYPE: int
                  DecList (5)
                    Dec (5)
                      VarDec (5)
                        ID: found
                      ASSIGNOP
                      Exp (5)
                        INT: 0
                  SEMI
        StmtList (6)
          Stmt (6)
            Exp (6)
              Exp (6)
                ID: a
              ASSIGNOP
              Exp (6)
                ID: read
                LP
                RP
            SEMI
          StmtList (7)
            Stmt (7)
              Exp (7)
                Exp (7)
                  ID: b
                ASSIGNOP
                Exp (7)
                  ID: read
                  LP
                  RP
              SEMI
            StmtList (8)
              Stmt (8)
                IF
                LP
                Exp (8)
                  Exp (8)
                    ID: a
                  RELOP
                  Exp (8)
                    ID: b
                RP
                Stmt (8)
                  CompSt (8)
                    LC
                    StmtList (9)
                      Stmt (9)
                        Exp (9)
                          Exp (9)
                            ID: start
                          ASSIGNOP
                          Exp (9)
                            ID: a
                        SEMI
                    RC
                ELSE
                Stmt (10)
                  CompSt (10)
                    LC
                    StmtList (11)
                      Stmt (11)
                        Exp (11)
                          Exp (11)
                            ID: start
                          ASSIGNOP
                          Exp (11)
                            ID: b
                        SEMI
                    RC
              StmtList (13)
                Stmt (13)
                  WHILE
                  LP
                  Exp (13)
                    Exp (13)
                      ID: found
                    RELOP
                    Exp (13)
                      INT: 0
                  RP
                  Stmt (13)
                    CompSt (13)
                      LC
                      StmtList (14)
                        Stmt (14)
                          IF
                          LP
                          Exp (14)
                            Exp (14)
                              ID: start
                            RELOP
                            Exp (14)
                              Exp (14)
                                LP
                                Exp (14)
                                  Exp (14)
                                    ID: start
                                  DIV
                                  Exp (14)
                                    ID: a
                                RP
                              STAR
                              Exp (14)
                                ID: a
                          RP
                          Stmt (14)
                            CompSt (14)
                              LC
                              StmtList (15)
                                Stmt (15)
                                  IF
                                  LP
                                  Exp (15)
                                    Exp (15)
                                      ID: start
                                    RELOP
                                    Exp (15)
                                      Exp (15)
                                        LP
                                        Exp (15)
                                          Exp (15)
                                            ID: start
                                          DIV
                                          Exp (15)
                                            ID: b
                                        RP
                                      STAR
                                      Exp (15)
                                        ID: b
                                  RP
                                  Stmt (15)
                                    CompSt (15)
                                      LC
                                      StmtList (16)
                                        Stmt (16)
                                          Exp (16)
                                            Exp (16)
                                              ID: result
                                            ASSIGNOP
                                            Exp (16)
                                              ID: start
                                          SEMI
                                        StmtList (17)
                                          Stmt (17)
                                            Exp (17)
                                              Exp (17)
                                                ID: found
                                              ASSIGNOP
                                              Exp (17)
                                                INT: 1
                                            SEMI
                                      RC
                                  ELSE
                                  Stmt (18)
                                    CompSt (18)
                                      LC
                                      StmtList (19)
                                        Stmt (19)
                                          Exp (19)
                                            Exp (19)
                                              ID: start
                                            ASSIGNOP
                                            Exp (19)
                                              Exp (19)
                                                ID: start
                                              PLUS
                                              Exp (19)
                                                INT: 1
                                          SEMI
                                      RC
                              RC
                          ELSE
                          Stmt (21)
                            CompSt (21)
                              LC
                              StmtList (22)
                                Stmt (22)
                                  Exp (22)
                                    Exp (22)
                                      ID: start
                                    ASSIGNOP
                                    Exp (22)
                                      Exp (22)
                                        ID: start
                                      PLUS
                                      Exp (22)
                                        INT: 1
                                  SEMI
                              RC
                      RC
                StmtList (25)
                  Stmt (25)
                    Exp (25)
                      ID: write
                      LP
                      Args (25)
                        Exp (25)
                          ID: result
                      RP
                    SEMI
                  StmtList (26)
                    Stmt (26)
                      RETURN
                      Exp (26)
                        INT: 0
                      SEMI
        RC
